File: leetcode/leetcode_cs/next-closest-time.cs
// https://leetcode.com/problems/next-closest-time/

public class Solution {
    public string NextClosestTime(string time) {
        var h = new HashSet<int>();
        foreach (var c in time) if (c != ':' && !h.Contains(c)) h.Add(c - '0');
        
        int start = Int32.Parse(time.Substring(0,2)) * 60 + Int32.Parse(time.Substring(3,2));
        int elpased = 60*24;
        while (true) {
            start = (start + 1) % (60*24);
            int[] times = {start/60/10, start/60%10, start%60/10, start%60%10};
            
            bool valid = true;
            foreach (var t in times) {
                if (!h.Contains(t)) {
                    valid = false;
                    break;
                }
            }
            if (!valid) continue;
            
            break;
        }
        
        return (start/60/10).ToString() + (start/60%10).ToString() + ":" + (start%60/10).ToString() + (start%60%10).ToString();
    }
}

File: leetcode/leetcode_py/next-closest-time.py
"""
https://leetcode.com/problems/next-closest-time/
"""

class Solution:
    def nextClosestTime_bruteforce(self, time: str) -> str:
        h = {int(x) for x in time if x != ':'}        
        start = int(time[:2]) * 60 + int(time[3:])
        
        while True:
            start = (start + 1) % 1440
            times = [start//60//10, start//60%10, start%60//10, start%60%10]
            
            valid = True
            for n in times:
                if n not in h:
                    valid = False
                    break
            if not valid: continue
                
            break
            
        return str(start//60//10) + str(start//60%10) + ":" + str(start%60//10) + str(start%60%10)
        
    def nextClosestTime(self, time: str) -> str:
        return self.nextClosestTime_bruteforce(time)